### ISSUE TITLE
ym подключение модуля из внешнего файла

### ISSUE THREAD

**Issue by kompolom** [bem/bem-forum-content-ru#576]
Всем привет. Есть ли возможность указать ym что модуль находится во внешнем файле, так чтобы тот его загрузил?

**Comment by zxqfox** [comment]
https://github.com/bem/bem-core/blob/v2/common.blocks/jquery/jquery.js — так? Или что ты имеешь ввиду? ;)

**Comment by tadatuta** [comment]
Сам по себе `ym` загружать файлы не умеет, но в `bem-core` для этого есть [loader](https://ru.bem.info/libs/bem-core/current/desktop/loader/).

**Comment by zxqfox** [comment]
Ой, он теперь еще и бандлы загружает... :+1: ничоси

**Comment by kompolom** [comment]
@zxqfox Да. В общем я выпилил из `_index.js` bh шаблоны в файл `index.bh.js`. Теперь из надо адекватно подключить.

**Comment by kompolom** [comment]
Кстати, о loader. Можно где то посмотреть реальный пример использования `loader_type_bundle`? штука интересная, но куда применить не знаю

**Comment by kompolom** [comment]
В отличие от jquery, где модуль собранный в один файл со всем остальным подгружает библиотеку, а затем провайдит ее, я пытаюсь загрузить сам модуль.

**Comment by kompolom** [comment]
Товарищи разработчики, подскажите что нибудь, пожалуйста.

**Comment by tadatuta** [comment]
@kompolom пример с `loader_type_bundle` когда-нибудь доберемся добавить в документацию, а по сути основного вопроса, будет выглядеть примерно так (не тестировал):
```js
modules.define('b1', ['i-bem__dom', 'loader_type_js'], function(provide, BEMDOM, Loader) {

provide(BEMDOM.decl(this.name, {
    onSetMod: {
        js: {
            inited: function() {
                var domElem = this.domElem;

                modules.isDefined('BH') || Loader('index.bh.js', function() {
                    modules.require('BH', function(bh) {
                        BEMDOM.update(domElem, bh.apply({ block: 'b2' }));
                    });
                });
            }
        }
    }
}));

});
```

**Comment by kompolom** [comment]
@tadatuta Получается, это придется писать во всех блоках использующих bh?

**Comment by zxqfox** [comment]
@kompolom Ну смотри. В _index.js тебе нужен какой-то код, который знает как загрузить свой файл с шаблонами. При чем, для каждого бандла нам нужен такой код, который будет загружать свой файл с шаблонами.

Почему не добавить в _index.js код типа:
```js
modules.define('BH', ['loader_type_js'], function (provide, Loader) {
  Loader('файл-бандла-с-бш.js', provide);
});
```

И дальше останется только загрузить его:
```js
modules.define('b1', ['i-bem__dom', 'BH'], function(provide, BEMDOM, bh) {
  // ...
```

Надо будет только придумать, как добавить в _index.js этот кусок кода с правильным путем до бандла. Устроит?

Устроит?

**Comment by kompolom** [comment]
@zxqfox Да, это уже интересней, я пробовал так сделать. В таком варианте bh провайдит `undefined`

**Comment by zxqfox** [comment]
А файло с bh.js что экспротирует? Попробуй руками в `<script>` его прописать и поймать то, что оно экспортировало. И руками Loader запустить: `modules.require('loader_type_js', function(Loader) { window.__Loader = Loader; или debugger; });`

**Comment by kompolom** [comment]
@zxqfox файл с bh.js экспортирует `bh` и `bemhtml`

**Comment by zxqfox** [comment]
https://github.com/bem/bem-core/blob/9fc75476e99c701a403e7be96b22922a4fb97deb/common.blocks/loader/_type/loader_type_js.js#L15

Надо так, потому что success не передает ничего:
```
modules.define('BH', ['loader_type_js'], function (provide, Loader) {
  Loader('файл-бандла-с-бш.js', function () {
    provide(window.BH); // поправь тут на правильное.
  });
});
```

**Comment by kompolom** [comment]
````
modules.define('BEMHTML',['loader_type_js'], function(provide, Loader) {
Loader('createpost.browser.bh.js', function(){
    modules.require('BH', function(bh){
        provide(bh);
    });
});

});
````
Так работает.
Но если делать modules.require('BEMHTML',....  то уже не работает. Похоже ym зацикливается.
А хотелось бы конечно максимально прозрачно... Например, чтобы модуль сам себя загружал и провайдил.

**Comment by zxqfox** [comment]
Толи лыжи не едут, толи... А зачем modules.require внутри provide внутри modules.define? ;-) Потому что в bh.js modules.define ?

Ну там внутри можно магию какую-то сделать на modules.isDefined и setTimeout'ах, кажется, что должно хватит ;-\. Либо через вспомогательный модуль, да. Каким является BEMHTML сейчас.

**Comment by kompolom** [comment]
@zxqfox внутри чего магию делать? Блока который грузим?

**Comment by zxqfox** [comment]
@kompolom заглушки этой, которая грузИТ.

Сейчас там:
```
modules.define('BEMHTML',['loader_type_js'], function(provide, Loader) {
Loader('merged.browser.bh.js', function(){
    modules.require('BH', function(bh){
```
Имена разные, двойное проникновение. Это уже магия своего рода ;-)

**Comment by kompolom** [comment]
@zxqfox Хочется вообще прозрачно сделать, как с jquery, без магии. но оно ж не робит.

**Comment by zxqfox** [comment]
Для этого надо, чтобы bh глобально торчал при загрузке, как торчит jquery. А для этого, в свою очередь, сборку бх.жс надо поправить.

**Comment by kompolom** [comment]
Ок. с bh можем настроить сборку. Но Вопрос несколько шире - грузить *любой* модуль по url

**Comment by kompolom** [comment]
Наверно тут уже в самом ym что то мудрить

**Comment by zxqfox** [comment]
modules.define = кастомная? Можно обертку над loader

**Comment by tadatuta** [comment]
@kompolom 
> Вопрос несколько шире - грузить любой модуль по url

а откуда вообще такое желание? `jQuery` мы так грузим с CDN, т.к. он на портале Яндекса по этим же урлам отдается и со значительной долей вероятности уже закеширован у пользователя в браузере. если же так делать с каждый модулем, то накладные расходы на загрузку будет очень высокими.

**Comment by kompolom** [comment]
@tadatuta Желание грузить то что пользователю может не понадобиться не сразу, а по мере необходимости.

**Comment by tadatuta** [comment]
Тогда это все-таки не все, а какие-то очень ограниченные куски и для них не должно быть проблемы написать кастомный код.
